New batch read endpoint

MatMoore opened the pull request:
This is a first pass at https://trello.com/c/DYFgXNYr/621-automate-link-checker-tool
Paired with @Davidslv and @benhyland 

This new endpoint allows us to page through content ids and retrieve all the content items
and links associated with that content id.

a response item looks like this:
```ruby
{
                  "content_id" => content_id,
                  "content_items" => [
                    {
                      "locale" => "en",
                      "base_path" => "/such-base-path",
                      "publishing_app" => "whitehall",
                      "format" => "guide",
                      "user_facing_version" => "2",
                      "state" => "draft",
                    },
                    {
                      "locale" => "en",
                      "base_path" => "/such-base-path",
                      "publishing_app" => "whitehall",
                      "format" => "guide",
                      "user_facing_version" => "1",
                      "state" => "published",
                    },
                  ],
                  "links" => {
                    "topics" => [topic_content_id],
                  }
}
```

## Motivation
We are planning to use this in tooling that will check consistency between
rummager and publishing api:
https://github.com/alphagov/finding-things-migration-checker

Content is currently indexed to rummager from multiple sources, and we don't
have a reliable way of rebuilding the index from scratch at the moment. In the
future, content will only be indexed from publishing api.  We want to ensure
that data is kept in sync until then.

We are also in the middle of a data migration from the old tagging
infrastructure (panopticon/content API) to the new one (publishing api/content
store).  Rummager is currently the most complete source of tagging data for
historical content, so comparing publishing API to that gives us a view on any
remaining discrepencies.

## Data returned
For now, we don't apply any filtering (e.g. by publication state). We also
don't want to expand the links hash to include things like base paths, since
this requires more information to uniquely identify the target content item
(such as locale) and we don't actually need this for our usage.

The fields returned for content items are based on the things that influence
content item uniqueness (see doc/object-model-explanation.md), plus
publishing-app and format. We could add more details, but we don't need this
now either.

The data is retrieved using 3 queries. A content ID has many content items and
many links. We query these separately because otherwise we would get a
cartesian product of rows (# of links x # of content items for each content
id).

The pagination works differently to other requests because content id does not
have an inherent ordering. This means that if we define a page by a limit and
and an offset, somebody could publish new content that appears in the middle of
our page between successive queries.

To ensure that the queries links and content items match up to the same set of
content ids, we query all the content ids on the page first, and then pass
those to the other queries.

## Suggestions for code review

We'd like at least one person from the publishing platform team to check the approach we've taken. 

The main commit adds a controller, a query, and a presenter.

The query itself runs all three queries and returns the results grouped together into an array of hashes.

We chose to mixin ActiveModel validations here as we had some fairly standard input validation to do and this seemed like a simple way of managing it. If there is a better way to do this please let us know. We're aware that the format of `render json: { errors: query.errors.full_messages }, status: 422` isn't quite correct for the publishing api but we wanted some feedback before we fix this.

We're using raw SQL because we thought the queries might get complicated, but it seems like we could do the same queries using activerecord too. Do you have a preference one way or the other?

The presenter takes the array returned by the query and returns something that can be returned as JSON to the client. We didn't use ResultPresenter because we are doing pagination a little differently. Passing in the last seen content id for pagination guarantees that if the requester pages through everything, they will never see the same item twice.

## Tests
While testing this I noticed a small issue that breaks CI sometimes:
 https://ci.dev.publishing.service.gov.uk/view/Publishing/job/govuk_publishing-api/303/console

The 2nd commit makes the expectation less strict. I can split this off into a separate PR if you want.

MatMoore commented:
cc @alphagov/team-finding-things

tijmenb commented on app/controllers/v2/grouped_content_and_links_controller.rb:
```diff
@@ -0,0 +1,26 @@
+  #
+  # This controller provides endpoints for querying content and links together.
+  #
+  class GroupedContentAndLinksController < ApplicationController
```
Should we name this closer to what we call it? `BatchReadController`?

MatMoore replied:
I'm not sure - I don't think either name describes this very well. Does anyone have any other suggestions?

tijmenb commented on app/queries/get_grouped_content_and_links.rb:
```diff
@@ -0,0 +1,116 @@
+module Queries
+  class GetGroupedContentAndLinks
```
Would it be interesting to explain the reason for using the `content_id` in pagination here?

MatMoore replied:
I've added a comment further down that explains why pagination works the way it does.

tijmenb commented on config/routes.rb:
```diff
@@ -27,6 +27,8 @@
 
       get "/linkables", to: "content_items#linkables"
+
+      get "/grouped_content_and_links", to: "grouped_content_and_links#index"
```
I think the pattern is to use dashes in paths.

MatMoore replied:
Yep. Actually we had trouble coming up with a descriptive name at all, so I'm interested if anyone has suggestions for something better to call this endpoint.

I don't think we have established language for talking about the thing that the content ID identifies - it's not something that's explicitly named in the data model at the moment. Other apps have a Document/Edition distinction but publishing api just has groups of Content Items that relate to each other.

tijmenb commented on spec/presenters/queries/grouped_content_and_links_spec.rb:
```diff
@@ -0,0 +1,93 @@
+require "rails_helper"
+
+module Presenters
+  module Queries
```
You can do `RSpec.describe Presenters:: Queries::GroupedContentAndLinks` so it doesn't need the deep nesting.

MatMoore replied:
Thanks, I've fixed this now.

tijmenb commented on spec/requests/grouped_content_and_links_spec.rb:
```diff
@@ -0,0 +1,153 @@
+private
+
+  def results(response)
+    JSON.parse(response.body, symbolize_names: true)[:results]
```
You can replace this with `parsed_respons["results"]` (https://github.com/alphagov/publishing-api/pull/299)

MatMoore replied:
Fixed

tijmenb commented on spec/queries/get_grouped_content_and_links_spec.rb:
```diff
@@ -0,0 +1,211 @@
+      end
+
+      context "with no links" do
+        it "returns the content item with empty links" do
```
This may be testing more than the description - perhaps you could just assert `expect(result["links"]).to eq([])` here, since the presence of the content is already tested in the request tests? Or perhaps split it up somehow?

MatMoore replied:
I think splitting it up into two tests would work. Will take a look.

MatMoore replied:
I've split the other assertions out into different tests. There's a bit of overlap but I think it's useful to test presence of content here too as it helps narrow down the source of a bug if you can rule out the presentation layer.

dougdroper commented:
I am curious about the pagination here, it seems to add a lot of incidental complexity to the implementation. 
I had a look at one of the content_items that has about 76 versions, and ran this query against it:
```sql
select content_items.id, content_items.format, content_items.content_id, translations.locale, locations.base_path, states.name, lock_versions.number from content_items
join locations on locations.content_item_id = content_items.id
join states on states.content_item_id = content_items.id
join lock_versions on lock_versions.target_id = content_items.id
join translations on translations.content_item_id = content_items.id
where content_items.content_id = '08d48cdd-6b50-43ff-a53b-beab47f4aab0'
group by content_items.id, translations.locale, locations.base_path, states.name, lock_versions.number
order by number DESC
```
and that takes about 0.2ms, so might be able to get away without pagination. 

Also this kinda feels like it is more like `Versions` rather then `GroupedContentAndLinks` so might need some thinking about how other people could use it too (specifically around the fields that would be displayed)

What do you think?

benhyland commented:
@dougdroper 

Hi Doug, I just want to explain a bit more about the motivation for the pagination here, since we might be understanding different things by the term.

The new endpoint is intended to be used for bulk reading, in batches of (say) 1000 content_ids per request.

We organise the response data under a hierarchy keyed by content_id because the endpoint is intended to deliver data which only makes sense per-content_id, such as links, in addition to the per-content_item_id data.

We therefore wish to paginate in pages of content_ids. In order to keep our pages stable in the face of edits that happen during an iteration across all pages, we must pass in the last_seen_content_id, rather than a numeric offset, since content_ids are not generated in sequence.

The result is that in a single iteration we can be sure of
- not missing a content_id that already existed when we started the first page
- not seeing any content_id twice
- seeing complete and consistent data for each content_id returned

If we iterate by some other key, such as content_item_id, or by pages using an offset rather than last_seen_content_id, these guarantees cannot be made.

Hope that makes sense - we can come and chat in person if you like?

benhyland commented:
@dougdroper regarding the naming and more general use cases, we tried to ensure that the endpoint could be useful to others as well, but we wanted to gather feedback before adding more fields or filtering.
Any suggestions (especially of specific use cases) are very welcome. Thanks!

dougdroper commented:
Ah right I see, so the response would something like:

```
{
  'total': 500000,
  'page': 1,
  ...
  'results': {
   [
     { 'content_id': '123', 'content_items': [{"base_path" => "/foo"}], 'links': {} },
     { 'content_id': '456', 'content_items': [{"base_path" => "/bar"}], 'links': {} }
   ]
  }
}
```

MatMoore commented:
@dougdroper Yep, that's right, except rather than `page` we return `last_seen_content_id`, for the reasons @benhyland  mentioned above.

We don't return `total` at the moment, but we can add that it in if it's going to be useful. We seem to be missing a couple of specs for the pagination response fields, so we'll need to take a look at that.

kaplan9000 commented on app/queries/get_grouped_content_and_links.rb:
```diff
@@ -0,0 +1,116 @@
+        JOIN
+          links on link_sets.id = links.link_set_id
+
+        WHERE link_sets.content_id IN (#{sql_value_placeholders(content_ids)})
```
`sql_value_placeholders` might never be anything dangerous right now but I think that adding an extra level of security around this query will not hurt and will prevent SQL injection in the case that in the future the code will change but the query will remain the same. 

Sadly, I appear not to find an easy way to escape SQL in multiline strings.

jscott06 replied:
sql_value_placeholders might never be anything dangerous right now but I think that adding an extra level of security around this query will not hurt and will prevent SQL injection in the case that in the future the code will change but the query will remain the same.

Sadly, I appear not to find an easy way to escape SQL in multiline strings.

jackscotti replied:
same as above

jackscotti commented on app/queries/get_grouped_content_and_links.rb:
```diff
@@ -0,0 +1,116 @@
+        JOIN states s on s.content_item_id = ci.id
+        JOIN user_facing_versions ufv on ufv.content_item_id = ci.id
+
+        WHERE ci.content_id IN (#{sql_value_placeholders(content_ids)})
```
sql_value_placeholders might never be anything dangerous right now but I think that adding an extra level of security around this query will not hurt and will prevent SQL injection in the case that in the future the code will change but the query will remain the same.

Sadly, I appear not to find an easy way to escape SQL in multiline strings.

dougdroper replied:
http://api.rubyonrails.org/classes/ActiveRecord/Sanitization/ClassMethods.html#method-i-sanitize_sql_array

MatMoore replied:
We're not trying to inject anything into the query here, only build up a dynamic list of placeholders like `IN ($1, $2, $3)`. In the `exec` call we pass the SQL and the values to the database separately, so this should be safer than trying to sanitise strings.

I think this query could be rewritten using `ContentItemFilter` though, which might make the code more readable. Any thoughts?

jackscotti commented on .gitignore:
```diff
@@ -16,3 +16,4 @@
 
 # Ignore the generated pact
 spec/pacts/publishing_api-content_store.json
+.DS_Store
```
I have added this to the `gitignore` in the past and I've been told that as because not every contributor uses a machine running Apple OS X, the `.DS_Store` should be ignored locally in the global `gitignore` file. 

http://technicalpickles.com/posts/configure-git-to-globally-ignore-some-files/

Davidslv replied:
There's nothing wrong in adding it to the repo, it should be there to avoid the mistake of adding a file that shouldn't be in the repo in the first place. If someone doesn't use a OSX machine that will have no impact in their workflow whatsoever.

jackscotti commented on app/queries/get_grouped_content_and_links.rb:
```diff
@@ -0,0 +1,116 @@
+    def initialize(last_seen_content_id: nil, page_size: nil)
+      @last_seen_content_id = last_seen_content_id || DEFAULT_CONTENT_ID
+      @page_size = page_size || PAGE_SIZE
+    end
```
```ruby
def initialize(last_seen_content_id: DEFAULT_CONTENT_ID, page_size: DEFAULT_PAGE_SIZE)
   @last_seen_content_id = last_seen_content_id
   @page_size = page_size
end
```

Would this be more readable and work the same way?

benhyland replied:
These are semantically different. With your proposed change, it is possible to pass and assign nil. In the original it is not.

dougdroper replied:
I believe if you call with `last_seen_content_id: nil`, it would override `DEFAULT_CONTENT_ID` with `nil` sadly

jackscotti replied:
oh. Learnt something new, thanks!

dougdroper commented:
What is the rational around returning all versions of the content_items?

benhyland commented:
@dougdroper Only that we haven't yet decided if or how we wish to filter them.

As an aside, at the moment we think it will be more flexible and maintainable for the future if we retrieve all versions from the db and filter in the application layer. The tradeoff is that the potentially unbounded versions in the result set might cause perf/memory issues. We don't expect this and we haven't looked into other potential mitigations such as streaming the response.

dougdroper commented on app/presenters/queries/grouped_content_and_links.rb:
```diff
@@ -0,0 +1,68 @@
+      end
+
+      def last_seen_content_id
+        return nil if results.empty?
```
the use of nil is redundant here
`return if results.empty?`

dougdroper commented on app/presenters/queries/grouped_content_and_links.rb:
```diff
@@ -0,0 +1,68 @@
+        groups = links.group_by { |link| link.fetch("link_type") }
+        links_hash = {}
+
+        groups.each do |link_type, type_links|
```
can use `each_with_object({})` to remove the links_hash variable

dougdroper commented on app/queries/get_grouped_content_and_links.rb:
```diff
@@ -0,0 +1,116 @@
+    end
+
+    def has_valid_uuid_for_pagination
+      unless @last_seen_content_id == DEFAULT_CONTENT_ID || UuidValidator.valid?(@last_seen_content_id)
```
Why do we need a UuidValidator? Maybe easier to see if a content_item exists with the content_id? Then we can drop the DEFAULT_CONTENT_ID

MatMoore replied:
It just makes for a friendlier API.

If the request is made with an invalid UUID, then the client is doing something wrong, and we could tell them why.

In this case though it's inconsistent with the other requests so I'll take it out. This means we will return 404 in all cases.

dougdroper replied:
I meant rather that we could do `ContentItem.exists?(last_seen_content_id)` as the validation. But am fine with it not validating.

dougdroper commented on app/queries/get_grouped_content_and_links.rb:
```diff
@@ -0,0 +1,116 @@
+
+      query = <<-SQL
+        SELECT
+          ci.content_id as content_id,
```
Can we use the names of the tables rather then the shortcut aliases, I think it is tricky to keep doing the conversion and helps with reading

dougdroper commented on app/queries/get_grouped_content_and_links.rb:
```diff
@@ -0,0 +1,116 @@
+    attr_accessor :last_seen_content_id, :page_size
+
+    validate :has_valid_uuid_for_pagination
+    validates_numericality_of :page_size, less_than_or_equal_to: MAX_PAGE_SIZE # FIXME: > 0
```
What does the comment # FIXME: > 0 mean?

dougdroper commented on app/queries/get_grouped_content_and_links.rb:
```diff
@@ -0,0 +1,116 @@
+        .having("content_id > ?", [last_seen_content_id])
+        .limit(page_size)
+        .order(:content_id)
+        .map(&:content_id)
```
can we change this to `pluck(:content_id)`

dougdroper commented on app/queries/get_grouped_content_and_links.rb:
```diff
@@ -0,0 +1,116 @@
+      ContentItem
+        .select(:content_id)
+        .group(:content_id)
+        .having("content_id > ?", [last_seen_content_id])
```
why is this an array?

dougdroper commented on app/queries/get_grouped_content_and_links.rb:
```diff
@@ -0,0 +1,116 @@
+    # array of content item rows, and an array of link rows.
+    def call
+      content_ids = query_content_ids_for_page(
+        last_seen_content_id: @last_seen_content_id,
```
we have an attr_reader for last_seen_content_id and page_size

MatMoore commented:
@dougdroper @tijmenb I've addressed your comments in a new commit.

If there are no other issues that need addressing now I'll rebase everything on top of the latest master (some integration tests are failing at the moment, but it seems unrelated to this work)

To elaborate on what we've discussed offline:

- We plan to run our checker tool only on integration for the time being
- @elliotcm is currently working on derived tables to improve read performance
- We don't actually need all versions of the content, although the state of the latest version is not enough for us to filter on. We think we should be able to rewrite the query to use Elliot's new tables when they're ready.
- We'd prefer to get this PR merged in the meantime, so we can start deploying the tool to integration.

I've raised this ticket in our backlog specifically to address the remaining work https://trello.com/c/3qIYFDBI/634-finish-publishing-api-endpoint-for-link-checker-tool

We haven't come up with a good name for the endpoint yet - I think we should rename it when we remove the multiple versions.

elliotcm commented:
Based on conversations offline (mentioned above) it doesn't like we're going to resolve everything in this PR.

I'm OK with us merging this based on the agreement that [this ticket](https://trello.com/c/3qIYFDBI/634-finish-publishing-api-endpoint-for-link-checker-tool) will be used to take a fresh look at the needs for this endpoint and the ways we can provide a robust endpoint via some careful denormalisation.  A member of the publishing platform team will be available to help with the discussion/planning of that, with the finding things team taking on the bulk of the actual typing.

elliotcm commented:
Get this green and I'll merge it.

MatMoore commented:
@elliotcm thanks, should be ready to merge now.
